drivers/console/ipm_console.c

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <drivers/ipm.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ipm_console, CONFIG_IPM_LOG_LEVEL);
+
+struct device *ipm_dev;
+
+static int console_out(int c)
+{
+	static char buf[CONFIG_IPM_CONSOLE_LINE_BUF_LEN];
+	static size_t len;
+	int ret;
+
+	if (c != '\n' && len < sizeof(buf)) {
+		buf[len++] = c;
+		return c;
+	}
```

Comment:
Wouldn't this miss the last out character if buffer is full?

1. Fill buffer to full. Now `len == sizeof(buf)` before `return`.
2. `console_out(c)` is called again with non-newline character.
3. Since `len == sizeof(buf)`, the assignment is skipped, and `ipm_send()` is called. But the character is not stored back into the buffer, and is not being sent thru IPM either. So this new character is lost.

Reply:
Thanks for good catch, fixed